packages/x-components/src/wiring/__tests__/debounce-operator.spec.ts

```diff
@@ -0,0 +1,195 @@
+         immediately if a previous forceOn event was emitted. */
+      bus.emit('UserClickedAResult', createResultStub('Random result'));
+      registerWire(debounce(wire, 500, { forceOn: 'UserClickedAResult' }));
+      emitWireEvent('1', '2');
+      expect(callback).not.toHaveBeenCalled();
+
+      bus.emit('UserClickedAResult', createResultStub('Random result'));
+      expect(callback).toHaveBeenCalledTimes(1);
+      expect(callback).toHaveBeenCalledWith('2');
+
+      jest.advanceTimersByTime(500);
+      expect(callback).toHaveBeenCalledTimes(1);
+
+      emitWireEvent('3', '4');
+      expect(callback).toHaveBeenCalledTimes(1);
+      jest.advanceTimersByTime(500);
+      expect(callback).toHaveBeenCalledTimes(2);
+    });
+
+    // eslint-disable-next-line max-len
+    it('executes the effect immediately if any one of the forceOn events is emitted while the debounce is running', () => {
+      const { wire, callback, emitWireEvent, registerWire, bus } = createWire();
+      registerWire(debounce(wire, 500, { forceOn: ['UserClickedAResult', 'UserScrolled'] }));
+      emitWireEvent('1', '2');
+      expect(callback).not.toHaveBeenCalled();
+
+      bus.emit('UserClickedAResult', createResultStub('Random result'));
+      expect(callback).toHaveBeenCalledTimes(1);
+      expect(callback).toHaveBeenCalledWith('2');
+
+      emitWireEvent('3', '4');
+      expect(callback).toHaveBeenCalledTimes(1);
+
+      bus.emit('UserScrolled', 100);
+      expect(callback).toHaveBeenCalledTimes(2);
+      expect(callback).toHaveBeenCalledWith('4');
+
+      jest.advanceTimersByTime(500);
+      expect(callback).toHaveBeenCalledTimes(2);
+    });
+  });
+
+  describe('when both cancel and force events are provided', () => {
+    it('ignores the cancel event if it is emitted after the force event', () => {
+      const { wire, callback, emitWireEvent, registerWire, bus } = createWire();
+      registerWire(
+        debounce(wire, 500, { forceOn: 'UserClickedAResult', cancelOn: 'UserClearedQuery' })
+      );
+      emitWireEvent('1', '2');
+      expect(callback).not.toHaveBeenCalled();
+
+      bus.emit('UserClickedAResult', createResultStub('Random result'));
+      bus.emit('UserClearedQuery', '');
+      expect(callback).toHaveBeenCalledTimes(1);
+      expect(callback).toHaveBeenCalledWith('2');
+
+      jest.advanceTimersByTime(500);
+      expect(callback).toHaveBeenCalledTimes(1);
+    });
+
+    it('does not emit when the force event is emitted after cancel', () => {
+      const { wire, callback, emitWireEvent, registerWire, bus } = createWire();
+      registerWire(
+        debounce(wire, 500, { forceOn: 'UserClickedAResult', cancelOn: 'UserClearedQuery' })
+      );
+      emitWireEvent('1', '2');
+      expect(callback).not.toHaveBeenCalled();
+
+      bus.emit('UserClearedQuery', '');
+      bus.emit('UserClickedAResult', createResultStub('Random result'));
+      expect(callback).toHaveBeenCalledTimes(0);
```

Comment:
Is there differences between .not.toHaveBeenCalled(); and .toHaveBeenCalledTimes(0);?

Reply:
The programmer's lack of care 🤣. I think `.not.toHaveBeenCalled()` is more readable